src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -677,6 +742,234 @@ impl Crumbable for crate::Block<Ast> {
+                Some(pat) => (val,pat),
+                None      => return missing("prefix"),
+            },
+            MatchCrumb::Segs{val,index} => match &val {
+                SegmentMatchCrumb::Body{val} => (val,&self.segs.get_or_err(*index,"elem")?.body),
+                SegmentMatchCrumb::Head      => return Ok(&self.segs.get_or_err(*index,"elem")?.head),
+            }
+        };
+
+        for crumb in path {
+            match (pattern.deref(),crumb) {
+                (Begin(_)    , PatternMatchCrumb::Begin   ) => return missing("elem"),
+                (End(_)      , PatternMatchCrumb::End     ) => return missing("elem"),
+                (Nothing(_)  , PatternMatchCrumb::Nothing ) => return missing("elem"),
+                (Build(pat)  , PatternMatchCrumb::Build   ) => return Ok(&pat.elem.wrapped),
+                (Err(pat)    , PatternMatchCrumb::Err     ) => return Ok(&pat.elem.wrapped),
+                (Tok(pat)    , PatternMatchCrumb::Tok     ) => return Ok(&pat.elem.wrapped),
+                (Blank(pat)  , PatternMatchCrumb::Blank   ) => return Ok(&pat.elem.wrapped),
+                (Var(pat)    , PatternMatchCrumb::Var     ) => return Ok(&pat.elem.wrapped),
+                (Cons(pat)   , PatternMatchCrumb::Cons    ) => return Ok(&pat.elem.wrapped),
+                (Opr(pat)    , PatternMatchCrumb::Opr     ) => return Ok(&pat.elem.wrapped),
+                (Mod(pat)    , PatternMatchCrumb::Mod     ) => return Ok(&pat.elem.wrapped),
+                (Num(pat)    , PatternMatchCrumb::Num     ) => return Ok(&pat.elem.wrapped),
+                (Text(pat)   , PatternMatchCrumb::Text    ) => return Ok(&pat.elem.wrapped),
+                (Block(pat)  , PatternMatchCrumb::Block   ) => return Ok(&pat.elem.wrapped),
+                (Macro(pat)  , PatternMatchCrumb::Macro   ) => return Ok(&pat.elem.wrapped),
+                (Invalid(pat), PatternMatchCrumb::Invalid ) => return Ok(&pat.elem.wrapped),
+                (Except(pat) , PatternMatchCrumb::Except  ) => pattern = &pat.elem,
+                (Tag(pat)    , PatternMatchCrumb::Tag     ) => pattern = &pat.elem,
+                (Cls(pat)    , PatternMatchCrumb::Cls     ) => pattern = &pat.elem,
+                (Or(pat)     , PatternMatchCrumb::Or      ) => match &pat.elem {
+                    crate::Switch::Left (pat) => pattern = &pat.value,
+                    crate::Switch::Right(pat) => pattern = &pat.value,
+                }
+                (Seq(pat), PatternMatchCrumb::Seq{right}) =>
+                    pattern = if *right {&pat.elem.1} else {&pat.elem.0},
+                (Many(pat), PatternMatchCrumb::Many{index}) =>
+                    pattern = pat.elem.get_or_err(*index,"elem")?,
+                _ => return missing("crumb"),
+
+            }
+        }
+        missing("elem")
+    }
+
+    fn set(&self, crumb:&Self::Crumb, new_ast:Ast) -> FallibleResult<Self> where Self: Sized {
+
+        fn set_pattern
```

Comment:
I'd place this in the root scope. It is well-defined function, no reason to hide it here.
Same for `iter_pat` and `get_pat`.